internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go

```diff
@@ -30,14 +35,90 @@ func NewUpgrade(log *logger.Logger, coord *coordinator.Coordinator) *Upgrade {
-func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, _ acker.Acker) error {
+// Handle handles UPGRADE action.  Returns immediately and the actual
+// upgrade happens asynchronously.  This allows for downloads to
+// happen without blocking updates.  If multiple upgrades are sent
+// then we ack them all if there is an error, but only the first actually executes.
+// If successful, reboot does ACK and check-in.
+func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker) error {
 	h.log.Debugf("handlerUpgrade: action '%+v' received", a)
 	action, ok := a.(*fleetapi.ActionUpgrade)
 	if !ok {
 		return fmt.Errorf("invalid type, expected ActionUpgrade and received %T", a)
 	}
 
-	// always perform PGP checks
-	return h.coord.Upgrade(ctx, action.Version, action.SourceURI, action, false)
+	asyncCtx, runAsync := h.getAsyncContext(ctx, a, ack)
+	if !runAsync {
+		return nil
+	}
+	go func() {
+		h.log.Infof("starting upgrade to version %s in background", action.Version)
+		if err := h.coord.Upgrade(asyncCtx, action.Version, action.SourceURI, action, false); err != nil {
+
+			h.log.Errorf("upgrade to version %s failed: %v", action.Version, err)
+			// If context is cancelled in getAsyncContext, the actions are acked there
+			if !errors.Is(asyncCtx.Err(), context.Canceled) {
+				h.bkgMutex.Lock()
+				h.ackActions(asyncCtx, ack)
+				h.bkgMutex.Unlock()
+			}
+		}
+	}()
+	return nil
+}
+
+// ackActions Acks all the actions in bkgActions, and deletes entries from bkgActions.
+// User is responsible for obtaining and releasing bkgMutex lock
+func (h *Upgrade) ackActions(ctx context.Context, ack acker.Acker) {
+	for _, a := range h.bkgActions {
+		if err := ack.Ack(ctx, a); err != nil {
+			h.log.Errorf("ack of failed upgrade failed: %v", err)
+		}
+	}
+	h.bkgActions = nil
+	if err := ack.Commit(ctx); err != nil {
+		h.log.Errorf("commit of ack for failed upgrade failed: %v", err)
+	}
+}
+
+// getAsyncContext returns a cancelContext and whether or not to run the upgrade
+func (h *Upgrade) getAsyncContext(ctx context.Context, action fleetapi.Action, ack acker.Acker) (context.Context, bool) {
+	h.bkgMutex.Lock()
+	defer h.bkgMutex.Unlock()
+	// If no existing actions, run this one
+	if len(h.bkgActions) == 0 {
+		h.bkgActions = append(h.bkgActions, action)
+		c, cancel := context.WithCancel(ctx)
+		h.bkgCancel = cancel
+		return c, true
+	}
+	// If upgrade to same version, save action to ack when first upgrade completes
+	upgradeAction, ok := action.(*fleetapi.ActionUpgrade)
+	if !ok {
+		h.log.Errorf("invalid type, expected ActionUpgrade and received %T", action)
+		return nil, false
+	}
+	// only need to check first action since all actions must be upgrades to same version
+	bkgAction, ok := h.bkgActions[0].(*fleetapi.ActionUpgrade)
+	if !ok {
+		h.log.Errorf("invalid type, expected ActionUpgrade and received %T", action)
+		return nil, false
+	}
+	if upgradeAction.Version == bkgAction.Version {
```

Comment:
This looks good, should we add in comparing the SourceURI?

Reply:
sure. done.